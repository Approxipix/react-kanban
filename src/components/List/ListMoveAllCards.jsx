import React, { Component } from 'react';
import PropTypes from "prop-types";
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { moveAllCards } from '../../redux/listReducer/actions/actions';
import { SubmitButton } from '../BaseComponent';
import Select from '../Select/Select';
import styled from 'styled-components';

const Container = styled.div`
  margin-bottom: 1rem;
`;

class ListMoveAllCards extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newListId: props.listId || null,
    }
  }

  moveAllCards =() => {
    const { listId } = this.props;
    const { newListId } = this.state;
    if (listId !== newListId) {
      this.props.actions.moveAllCards({
        listId: listId,
        newListId: newListId,
      });
    }
    this.props.toggleOpened();
  };

  listIdSelector = () => {
    const { newListId } = this.state;
    const { boards, boardId, lists, listId } = this.props;
    const listsOptions = boards[boardId].lists.map((id) => ({
      title: listId === id ? `${lists[id].title} (current)` : lists[id].title,
      value: id,
    }));

    return (
      <Select
        label="Board"
        placeholder="Select board"
        value={lists[newListId].title}
        options={listsOptions}
        onChange={(id) => this.setState({ newListId: id })}
      />
    );
  };

  render() {
    return (
      <>
        <Container>
          {this.listIdSelector()}
        </Container>
        <SubmitButton onClick={() => this.moveAllCards()}>
          Move
        </SubmitButton>
      </>
    );
  }
}

ListMoveAllCards.defaultProps = {
  boards: null,
  lists: null,
};

ListMoveAllCards.propTypes = {
  boardId: PropTypes.string.isRequired,
  listId: PropTypes.string.isRequired,
  boards: PropTypes.objectOf(
    PropTypes.shape({
      _boardId: PropTypes.string,
      title: PropTypes.string,
      color: PropTypes.string,
      lists: PropTypes.arrayOf(PropTypes.string).isRequired,
    })
  ).isRequired,
  lists: PropTypes.objectOf(
    PropTypes.shape({
      _listId: PropTypes.string,
      title: PropTypes.string.isRequired,
      cards: PropTypes.arrayOf(PropTypes.string)
    })
  ).isRequired,
};

function mapStateToProps(state) {
  return {
    boards: state.boardReducer,
    lists: state.listReducer,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      moveAllCards: moveAllCards,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ListMoveAllCards);
