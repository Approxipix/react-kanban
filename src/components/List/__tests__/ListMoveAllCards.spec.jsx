import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListMoveAllCards from '../ListMoveAllCards';

const mockStore = configureStore([]);

describe('<ListMoveAllCards>', () => {
  let store;
  let props;
  let component;
  let mockToggleOpened= jest.fn();
  let mockMoveAllCards = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    store = mockStore({
      boardReducer: {
        Board1ID: {
          _boardId: 'Board1ID',
          title: 'Board 1',
          lists: ['List1ID', 'List2ID'],
        },
      },
      listReducer: {
        List1ID: {
          _listId: 'List1ID',
          title: 'List1 Title',
          cards: ['List1 Card1', 'List1 Card2']
        },
        List2ID: {
          _listId: 'List2ID',
          title: 'List2 Title',
          cards: ['List2 Card1', 'List2 Card2']
        },
      }
    });
    store.dispatch = mockMoveAllCards;
    props = {
      boardId: 'Board1ID',
      listId: 'List1ID',
      toggleOpened: mockToggleOpened
    };
    const state = {
      newListId: 'List2ID',
    };

    const wrapper = shallow(<ListMoveAllCards store={store} {...props} />);
    component = wrapper.find('ListMoveAllCards').dive();
    component.setState({ ...state });
  });

  it('should render with given state from Redux store and correct props', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call toggleOpened if listId/newListId are the same', () => {
    component.setState({ newListId: 'List1ID' });
    component.find('SubmitButton').simulate('click');

    expect(mockToggleOpened).toHaveBeenCalledTimes(1);
  });

  it("should dispatch an action if listId/newListId are different", () => {
    const expectedAction = {
      type: 'MOVE_ALL_CARDS',
      payload: {
        listId: 'List1ID',
        newListId: 'List2ID',
      }
    };

    component.find('SubmitButton').simulate('click');

    expect(mockMoveAllCards).toHaveBeenCalledTimes(1);
    expect(mockMoveAllCards).toHaveBeenCalledWith(expectedAction);
  });
});
