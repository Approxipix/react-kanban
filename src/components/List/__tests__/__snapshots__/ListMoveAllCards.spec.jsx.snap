// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListMoveAllCards> should render with given state from Redux store and correct props 1`] = `
<Fragment>
  <styled.div>
    <Select
      label="Board"
      onChange={[Function]}
      options={
        Array [
          Object {
            "title": "List1 Title (current)",
            "value": "List1ID",
          },
          Object {
            "title": "List2 Title",
            "value": "List2ID",
          },
        ]
      }
      placeholder="Select board"
      value="List2 Title"
    />
  </styled.div>
  <SubmitButton
    onClick={[Function]}
  >
    Move
  </SubmitButton>
</Fragment>
`;
